File: assignment-1-transformations/2/codes/myLinearContrastStretching.m
function img = myLinearContrastStretching(org)


org = im2double(org);
% Works for any aspect ratio
org_min = repmat(min(min(org)), size(org, 1), size(org, 2));
org_max = repmat(max(max(org)), size(org, 1), size(org, 2));
img = mat2gray((org - org_min) * 255 ./ (org_max - org_min));

%org = imread('../data/barbara.png');
%img = myLinearContrastStretching(org);

figure('name', 'LinearContrastStretching on Barbara')
colormap(jet(200));
subplot(2, 2, 1), imagesc(org);
title('Barbara')
colorbar;
subplot(2, 2, 2), imagesc(img);
title('After LCS')
colorbar;
subplot(2, 2, 3), imhist(org);
subplot(2, 2, 4), imhist(img);

File: assignment-1-transformations/2/codes/myLinearContrastStretching.py
import numpy as np
from matplotlib import pyplot as plt
from PIL import Image


def linearContrastStretch(img_path, mask=None):
    img = Image.open(img_path)
    img = np.array(img)

    if mask is not None:
        img = mask * img + (1 - mask)
        img_m = img.reshape(-1)
        img_m = img_m[np.where(img_m>=0)]
    else:
        img_m = img.reshape(img.shape[0]*img.shape[1], -1)


    img_min, img_max = img_m.min(0), img_m.max(0)
    curr_range = img_max - img_min
    img_streched = ((img - img_min)/curr_range)*255
    img_streched = img.astype(np.uint8)

    a,b = 1, 2
    if img.shape[0] > img.shape[1]*2:
        a,b = 2,1

    plt.subplot(a, b, 1)
    plt.imshow(img, cmap='gray')
    plt.title('Original Image')
    if img.ndim == 2:
        plt.colorbar()

    plt.subplot(a, b, 2)
    plt.imshow(img_streched, cmap='gray')
    plt.title('Contrast Stretched Image')
    if img.ndim == 2:
        plt.colorbar()
    plt.show()


if __name__ == "__main__":
    linearContrastStretch('../data/church.png')
